File: beam-android/core/ffi/src/main/kotlin/dev/beam/android/core/ffi/repository/Repositories.kt
package dev.beam.android.core.ffi.repository

import kotlinx.coroutines.flow.Flow
import uniffi.beam_client_core.AdminEvent
import uniffi.beam_client_core.AdminLogEntry
import uniffi.beam_client_core.AdminStatus
import uniffi.beam_client_core.AdminUserPage
import uniffi.beam_client_core.BrowseQuery
import uniffi.beam_client_core.ContinueWatchingEntry
import uniffi.beam_client_core.DeviceProfile
import uniffi.beam_client_core.DeviceSession
import uniffi.beam_client_core.EpisodeSummary
import uniffi.beam_client_core.HistoryPage
import uniffi.beam_client_core.LibraryFileSummary
import uniffi.beam_client_core.LibrarySummary
import uniffi.beam_client_core.MediaDetail
import uniffi.beam_client_core.MediaPage
import uniffi.beam_client_core.MediaSourceView
import uniffi.beam_client_core.PlaybackHttpConfig
import uniffi.beam_client_core.ProgressOutcome
import uniffi.beam_client_core.QualityPolicy
import uniffi.beam_client_core.ServerHealth
import uniffi.beam_client_core.ServerSummary
import uniffi.beam_client_core.SessionState
import uniffi.beam_client_core.SourceSelection
import uniffi.beam_client_core.UserSummary

// Every screen depends on one of these interfaces rather than on `BeamClient`
// directly. That is what lets a feature's tests run on the JVM with a fake,
// with no `.so` loaded and no JNA on the classpath -- which in turn is what
// makes the Android test suite runnable in CI without an emulator.

/** Which servers are known, and who we are on them. */
public interface ServerRepository {
    /** The known servers, re-emitted whenever one is added, chosen or removed. */
    public val servers: Flow<List<ServerSummary>>

    /** Load the registry and any stored sessions. Called once at app start. */
    public suspend fun restore(): List<ServerSummary>

    /** Add a server by address, and make it the active one. */
    public suspend fun addServer(baseUrl: String, displayName: String?): ServerSummary

    /** Make an already-known server the active one. */
    public suspend fun selectServer(serverId: String)

    /** Forget a server, its session, and its queued progress. */
    public suspend fun removeServer(serverId: String)

    /** The URL to open in the in-app browser to sign in. */
    public suspend fun loginUrl(serverId: String): String

    /** Hand over a cookie lifted from the browser, and verify it. */
    public suspend fun completeLogin(serverId: String, sessionCookie: String): UserSummary

    /** End the session on this device. */
    public suspend fun logout(serverId: String)

    /** The current authentication state of a server. */
    public suspend fun sessionState(serverId: String): SessionState

    /** The active server, or null when none is chosen. */
    public suspend fun activeServer(): ServerSummary?
}

/** Browsing the catalog. */
public interface CatalogRepository {
    /** One page of the catalog, filtered and sorted. */
    public suspend fun browse(query: BrowseQuery): MediaPage

    /** Everything a detail screen shows for one title. */
    public suspend fun detail(mediaId: String): MediaDetail

    /** Every genre in the catalog, for the filter chips. */
    public suspend fun genres(): List<String>

    /** Every library on the server. */
    public suspend fun libraries(): List<LibrarySummary>

    /** One library. */
    public suspend fun library(libraryId: String): LibrarySummary

    /** The files indexed into one library. */
    public suspend fun libraryFiles(libraryId: String): List<LibraryFileSummary>

    /** The next playable episode after this one, across season boundaries. */
    public suspend fun upNext(showId: String, currentEpisodeId: String): EpisodeSummary?
}

/** Choosing what to play, playing it, and recording where the viewer got to. */
public interface PlaybackRepository {
    /** Tell the core what this device can decode. */
    public suspend fun setDeviceProfile(profile: DeviceProfile)

    /** Every file behind a title, playable or not. */
    public suspend fun sources(mediaId: String): List<MediaSourceView>

    /** The file to play, plus why, plus why each other file was rejected. */
    public suspend fun selectSource(mediaId: String, policy: QualityPolicy): SourceSelection

    /** URL, headers and pins for the platform player to fetch bytes itself. */
    public suspend fun playbackConfig(fileId: String): PlaybackHttpConfig

    /** Partially-watched titles, ready to resume. */
    public suspend fun continueWatching(limit: UInt?): List<ContinueWatchingEntry>

    /** One page of watch history. */
    public suspend fun history(limit: UInt?, offset: UInt?): HistoryPage

    /** Report where the viewer is, subject to the shared throttle. */
    public suspend fun reportProgress(
        fileId: String,
        positionSecs: Double,
        durationSecs: Double?,
        force: Boolean,
    ): ProgressOutcome

    /** Send every queued position that is due. Returns how many landed. */
    public suspend fun flushProgress(): UInt

    /** How many positions are waiting to be sent. */
    public suspend fun pendingProgressCount(): UInt
}

/** The signed-in devices on the profile screen. */
public interface SessionRepository {
    /** Every device signed in as this user. */
    public suspend fun sessions(): List<DeviceSession>

    /** Revoke one device. */
    public suspend fun revoke(sessionId: String)

    /** End every session everywhere, including this one. */
    public suspend fun logoutEverywhere()
}

/** The administrative surface, available only to administrators. */
public interface AdminRepository {
    /** The dashboard snapshot. */
    public suspend fun status(): AdminStatus

    /** The server's own health report. */
    public suspend fun health(): ServerHealth

    /** One page of user accounts. */
    public suspend fun users(limit: UInt?, offset: UInt?): AdminUserPage

    /** Block or unblock an account. */
    public suspend fun setUserDisabled(userId: String, disabled: Boolean)

    /** One page of the operational log. */
    public suspend fun logs(limit: UInt?, offset: UInt?): List<AdminLogEntry>

    /** How many log lines the server holds. */
    public suspend fun logCount(): ULong

    /** Recent server events, newest first. */
    public suspend fun events(limit: UInt?): List<AdminEvent>

    /** Create a library from a path on the server. */
    public suspend fun createLibrary(name: String, rootPath: String): LibrarySummary

    /** Delete a library and everything indexed into it. */
    public suspend fun deleteLibrary(libraryId: String)

    /** Rescan a library. Returns how many files were added. */
    public suspend fun scanLibrary(libraryId: String): UInt

    /** Re-fetch metadata for one title. */
    public suspend fun refreshMetadata(mediaId: String)
}

File: beam-apple/Packages/BeamKit/Sources/BeamCore/Repositories.swift
import BeamFFI
import BeamModel
import Foundation

// The six seams every screen talks to, mirroring `Repositories.kt`.
//
// Screens depend on these protocols and never on `BeamClient`, so a view model
// can be driven by a stateful fake from `BeamTesting` with no server, no
// network and no session. Every method throws `BeamFailure` rather than the
// generated `BeamError`, so the "what do I show, can I retry, must they sign
// in" decision is made once here instead of in ten view models.

/// The registry of servers this device knows, and the session on each.
public protocol ServerRepository: Sendable {
    /// Load the registry from storage. Called once at launch.
    func restore() async throws -> [ServerSummary]
    /// Every known server.
    func listServers() throws -> [ServerSummary]
    /// Add a server, or return the existing entry for the same origin.
    func addServer(baseURL: String, displayName: String?) async throws -> ServerSummary
    /// Make `serverId` the one subsequent calls act on.
    func selectServer(id serverId: String) async throws
    /// Forget a server, its session and its trusted certificates.
    func removeServer(id serverId: String) async throws
    /// The URL to open in a web view to begin signing in.
    func loginURL(serverId: String) throws -> String
    /// Hand the cookie lifted from the web view to the core, which verifies it.
    func completeLogin(serverId: String, sessionCookie: String) async throws -> UserSummary
    /// End the session on one server.
    func logout(serverId: String) async throws
    /// The session state for one server.
    func sessionState(serverId: String) throws -> SessionState
    /// Accept one certificate, by whole-certificate SHA-256 fingerprint.
    func trustCertificate(serverId: String, fingerprint: String) async throws
    /// Drop every certificate accepted for a server.
    func forgetCertificates(serverId: String) async throws
    /// The fingerprints currently accepted for a server.
    func trustedCertificates(serverId: String) throws -> [String]
}

/// Browsing the catalogue.
public protocol CatalogRepository: Sendable {
    /// One page of the catalogue.
    func browse(query: BrowseQuery) async throws -> MediaPage
    /// One title, with its seasons and episodes where it is a show.
    func detail(mediaId: String) async throws -> MediaDetail
    /// Every genre present in the library, for the filter control.
    func genres() async throws -> [String]
    /// The next episode worth playing after `currentEpisodeId`.
    func upNext(showId: String, currentEpisodeId: String) async throws -> EpisodeSummary?
    /// Every library.
    func libraries() async throws -> [LibrarySummary]
    /// One library.
    func library(id libraryId: String) async throws -> LibrarySummary
    /// The files indexed into one library.
    func libraryFiles(libraryId: String) async throws -> [LibraryFileSummary]
}

/// Choosing what to play, playing it, and recording that it was played.
public protocol PlaybackRepository: Sendable {
    /// Tell the core what this device can decode. Called once at launch and
    /// again whenever the software-decode preference changes.
    func setDeviceProfile(_ profile: DeviceProfile)
    /// Every file behind a title, playable or not.
    func sources(mediaId: String) async throws -> [MediaSourceView]
    /// The core's pick among them, with the reasons the rest were rejected.
    func selectSource(mediaId: String, policy: QualityPolicy) async throws -> SourceSelection
    /// The URL, credential and pinned certificates the player needs.
    func playbackConfig(fileId: String) throws -> PlaybackHttpConfig
    /// The same, for a whole server rather than one file. Downloads need this.
    func serverHTTPConfig() throws -> ServerHttpConfig
    /// Report a playback position. The core throttles and queues.
    func reportProgress(
        fileId: String,
        positionSeconds: Double,
        durationSeconds: Double?,
        force: Bool
    ) async throws -> ProgressOutcome
    /// Send whatever the queue is holding.
    func flushProgress() async throws -> UInt32
    /// How much the queue is holding.
    func pendingProgressCount() async throws -> UInt32
    /// Titles this person is part way through.
    func continueWatching(limit: UInt32?) async throws -> [ContinueWatchingEntry]
    /// One page of watch history.
    func history(limit: UInt32?, offset: UInt32?) async throws -> HistoryPage
}

/// The signed-in person's own sessions across their devices.
public protocol SessionRepository: Sendable {
    /// Every active session.
    func sessions() async throws -> [DeviceSession]
    /// End one session, on any device.
    func revoke(sessionId: String) async throws
    /// End every session, including this one.
    func logoutEverywhere() async throws
}

/// Operator actions. Guarded by the server, not by hiding the screen: a
/// non-administrator gets a 403, which surfaces as `BeamFailure.isForbidden`.
public protocol AdminRepository: Sendable {
    /// The dashboard snapshot.
    func status() async throws -> AdminStatus
    /// One page of user accounts.
    func users(limit: UInt32?, offset: UInt32?) async throws -> AdminUserPage
    /// Disable or re-enable an account.
    func setUserDisabled(userId: String, disabled: Bool) async throws
    /// One page of the operational log.
    func logs(limit: UInt32?, offset: UInt32?) async throws -> [AdminLogEntry]
    /// How many log lines the server holds.
    func logCount() async throws -> UInt64
    /// Recent server events.
    func events(limit: UInt32?) async throws -> [AdminEvent]
    /// Add a library rooted at a path on the server.
    func createLibrary(name: String, rootPath: String) async throws -> LibrarySummary
    /// Remove a library.
    func deleteLibrary(id libraryId: String) async throws
    /// Scan a library, returning how many files were added.
    func scanLibrary(id libraryId: String) async throws -> UInt32
    /// Re-run enrichment for one title.
    func refreshMetadata(mediaId: String) async throws
    /// The server's own health report.
    func health() async throws -> ServerHealth
}

/// Local settings, which never leave the device.
public protocol PreferencesRepository: Sendable {
    /// The current preferences.
    func load() -> UserPreferences
    /// Replace them.
    func save(_ preferences: UserPreferences)
}
